OpenRA.Mods.Common/Activities/Attack.cs

```diff
@@ -186,9 +186,28 @@ protected virtual AttackStatus TickAttack(Actor self, AttackFrontal attack)
 			if (armaments.Count == 0)
 				return AttackStatus.UnableToAttack;
 
-			// Update ranges
-			minRange = armaments.Max(a => a.Weapon.MinRange);
-			maxRange = armaments.Min(a => a.MaxRange());
+			// Determine whether the actor needs to move before it can fire:
+			// * The first entry in Attack*.Info.Armaments defines the actor's "primary" armament(s),
+			//   which are used for determining the ideal attack range for the actor.
+			//   Additional entries are treated as secondary armaments which will be used if they are in
+			//   range, but do not influence the actor's movement.
+			// * If the actor can move, the ideal location is calculated as the intersection of the ranges
+			//   of the primary armaments. This ensures the actor is able to attack using all of its
+			//   primary armaments.
+			// * If the actor cannot move, the ideal location check simplifies to whether *any* primary
+			//   armament is able to fire, calculated using the union of ranges, since attacking using
+			//   a subset of weapons is better than sitting idle.
+			var primaryArmaments = armaments.Where(a => a.Info.Name == attack.Info.Armaments.First());
```

Comment:
`var armaments` only includes weapons valid against this target.

In cases like Mammoth tanks, only `secondary` might be valid, resulting in `primaryArmaments` being null.

Maybe just change the above to `FirstOrDefault` and fall back to `armaments` if `primaryArmaments` is still null?

Alternatively, we could work on the assumption that `primary` and `secondary` will remain the only levels used in our official mods and many 3rd-party mods, and do something like this to try to fall back to `secondary` before falling back to bleed behavior:
```
var armamentsToDetermineRange = armaments.Where(a => a.Info.Name == attack.Info.Armaments.FirstOrDefault());
if (armamentsToDetermineRange == null)
	armamentsToDetermineRange = armaments.Where(a => a.Info.Name == attack.Info.Armaments.Skip(1).FirstOrDefault());

if (armamentsToDetermineRange == null)
	armamentsToDetermineRange = armaments;
```